src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.h

```diff
@@ -0,0 +1,206 @@
+    bool kickDone(const Update& event);
+
+    /**
+     * Guard on whether the placing robot has finished placing the ball into the desired
+     * position
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     *
+     * @return whether the ball has been placed into the desired position
+     */
+    bool ballPlaced(const Update& event);
+
+    /**
+     * Guard on whether the robot has waited stationary for 5 seconds
+     * @param event the BallPlacementPlayFSM Update event
+     * @return whether the robot has waited for 5 seconds
+     */
+    bool waitDone(const Update& event);
+
+    /**
+     * Guard on whether the robot has retreated outside of the required range
+     * @param event the BallPlacementPlayFSM Update event
+     * @return whether the robot has retreated outside of the required range
+     */
+    bool retreatDone(const Update& event);
+
+    /**
+     * Helper function for calculating the angle to kick the ball off of a wall
+     *
+     * @param ball_pos the ball position to use when calculating the kick angle
+     * @param field_lines the field lines of the playing area
+     *
+     * @return the kick angle
+     */
+    Angle calculateWallKickoffAngle(const Point& ball_pos, const Rectangle& field_lines);
+
+    /**
+     * Helper function that populates the move_tactics field with MoveTactics that
+     * organize the robots away from the ball placing robot
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     */
+    void setupMoveTactics(const Update& event);
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        DEFINE_SML_STATE(StartState)
+        DEFINE_SML_STATE(KickOffWallState)
+        DEFINE_SML_STATE(AlignPlacementState)
+        DEFINE_SML_STATE(PlaceBallState)
+        DEFINE_SML_STATE(WaitState)
+        DEFINE_SML_STATE(RetreatState)
+        DEFINE_SML_EVENT(Update)
+
+        DEFINE_SML_ACTION(alignPlacement)
+        DEFINE_SML_ACTION(placeBall)
+        DEFINE_SML_ACTION(kickOffWall)
+        DEFINE_SML_ACTION(startWait)
+        DEFINE_SML_ACTION(retreat)
+
+        DEFINE_SML_GUARD(shouldKickOffWall)
+        DEFINE_SML_GUARD(alignDone)
+        DEFINE_SML_GUARD(kickDone)
+        DEFINE_SML_GUARD(ballPlaced)
+        DEFINE_SML_GUARD(waitDone)
+        DEFINE_SML_GUARD(retreatDone)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *StartState_S + Update_E[!shouldKickOffWall_G] / alignPlacement_A =
+                AlignPlacementState_S,
+            StartState_S + Update_E[shouldKickOffWall_G] = KickOffWallState_S,
+            KickOffWallState_S + Update_E[!kickDone_G && shouldKickOffWall_G] /
+                                     kickOffWall_A                = KickOffWallState_S,
+            KickOffWallState_S + Update_E[kickDone_G]             = KickOffWallState_S,
+            KickOffWallState_S + Update_E[!kickDone_G]            = AlignPlacementState_S,
```

Comment:
shouln't the logic of these two statements be flipped

Reply:
Its a little weird because kickdone basically checks that the ball is still in motion, so its basically checking if the ball isn't moving anymore before going to the next state